File: analysis/import/SonarImporter/src/main/java/de/maibornwolff/codecharta/importer/sonar/SonarMeasuresAPIImporter.java
package de.maibornwolff.codecharta.importer.sonar;

import de.maibornwolff.codecharta.importer.sonar.dataaccess.SonarMeasuresAPIDatasource;
import de.maibornwolff.codecharta.importer.sonar.dataaccess.SonarMetricsAPIDatasource;
import de.maibornwolff.codecharta.importer.sonar.model.ComponentMap;
import de.maibornwolff.codecharta.model.Project;
import de.maibornwolff.codecharta.translator.MetricNameTranslator;

import java.util.List;

public class SonarMeasuresAPIImporter {

    private final SonarMeasuresAPIDatasource measuresDS;
    private final SonarMetricsAPIDatasource metricsDS;
    private final SonarCodeURLLinker sonarCodeURLLinker;
    private final MetricNameTranslator translator;
    private final boolean usePath;

    public SonarMeasuresAPIImporter(SonarMeasuresAPIDatasource measuresDS, SonarMetricsAPIDatasource metricsDS) {
        this(measuresDS, metricsDS, SonarCodeURLLinker.NULL, MetricNameTranslator.TRIVIAL, false);
    }

    public SonarMeasuresAPIImporter(SonarMeasuresAPIDatasource measuresDS, SonarMetricsAPIDatasource metricsDS, SonarCodeURLLinker sonarCodeURLLinker, MetricNameTranslator translator, boolean usePath) {
        this.measuresDS = measuresDS;
        this.metricsDS = metricsDS;
        this.sonarCodeURLLinker = sonarCodeURLLinker;
        this.translator = translator;
        this.usePath = usePath;
    }

    public Project getProjectFromMeasureAPI(String projectKey, String projectName, List<String> metrics) {
        List<String> metricsList = getMetricList(metrics);
        ComponentMap componentMap = measuresDS.getComponentMap(projectKey, metricsList);

        SonarComponentProjectAdapter project = new SonarComponentProjectAdapter(projectName, sonarCodeURLLinker, translator, usePath);
        project.addComponentMapsAsNodes(componentMap);

        return project;
    }

    public List<String> getMetricList(List<String> metrics) {
        if (metrics.isEmpty()) {
            return metricsDS.getAvailableMetricKeys();
        }
        return metrics;
    }

}

File: analysis/import/SonarImporter/src/main/kotlin/de/maibornwolff/codecharta/importer/sonar/SonarMeasuresAPIImporter.kt
package de.maibornwolff.codecharta.importer.sonar

import de.maibornwolff.codecharta.importer.sonar.dataaccess.SonarMeasuresAPIDatasource
import de.maibornwolff.codecharta.importer.sonar.dataaccess.SonarMetricsAPIDatasource
import de.maibornwolff.codecharta.model.Project
import de.maibornwolff.codecharta.translator.MetricNameTranslator
import mu.KotlinLogging

private val logger = KotlinLogging.logger {}

class SonarMeasuresAPIImporter @JvmOverloads constructor(
    private val measuresDS: SonarMeasuresAPIDatasource?,
    private val metricsDS: SonarMetricsAPIDatasource?,
    private val sonarCodeURLLinker: SonarCodeURLLinker = SonarCodeURLLinker.NULL,
    private val translator: MetricNameTranslator = MetricNameTranslator.TRIVIAL,
    private val usePath: Boolean = false
) {

    fun getProjectFromMeasureAPI(projectKey: String, metrics: List<String>): Project {
        val metricsList = getMetricList(metrics)
        logger.info { "Get values for metrics $metricsList." }

        val componentMap = measuresDS!!.getComponentMap(projectKey, metricsList)

        val projectBuilder = SonarComponentProjectBuilder(sonarCodeURLLinker, translator, usePath)
        val attributeDescriptors = getAttributeDescriptors()
        projectBuilder.addAttributeDescriptions(
            attributeDescriptors
                .filter { it.key in metricsList }
                .map { translator.translate(it.key) to it.value }
                .toMap()
        )
        return projectBuilder.addComponentMapsAsNodes(componentMap).build()
    }

    fun getMetricList(metrics: List<String>): List<String> {
        return if (metrics.isEmpty()) {
            metricsDS!!.availableMetricKeys
        } else metrics
    }
}
